### PR TITLE
exchange/wrapper: add GetServerTime() for exchange analytics

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

Add GetServerTime() to exchange interface. 

Fixes/Starts #763 

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/interfaces.go [hunk 1]
```diff
@@ -98,6 +98,9 @@ type IBotExchange interface {
 	UpdateOrderExecutionLimits(ctx context.Context, a asset.Item) error
 
 	AccountManagement
+
+	// GetServerTime returns the current exchange server time.
+	GetServerTime(ctx context.Context, ai asset.Item) (time.Time, error)
```
Comment (gloriousCode) [inline exchanges/interfaces.go:103]: I don't think this doesn't need its own section and commentary, please add to the main chunk of functions.

File: exchanges/binance/binance_wrapper.go [hunk 2]
```diff
@@ -1882,3 +1882,15 @@ func (b *Binance) formatUSDTMarginedFuturesPair(p currency.Pair, pairFmt currenc
...
+// GetServerTime returns the current exchange server time.
+func (b *Binance) GetServerTime(ctx context.Context, ai asset.Item) (time.Time, error) {
+	if !ai.IsValid() {
+		return time.Time{}, fmt.Errorf("%s %w", ai, asset.ErrNotSupported)
+	}
+
+	if ai == asset.USDTMarginedFutures {
+		return b.UServerTime(ctx)
+	}
+	return time.Time{}, common.ErrNotYetImplemented
```
Comment (gloriousCode) [inline exchanges/binance/binance_wrapper.go:1895]: Server time is returned with:
Spot:
`GetExchangeInfo(...)`
CoinMarginedFutures:
`FuturesExchangeInfo(...)`

File: exchanges/btcmarkets/btcmarkets_wrapper.go [hunk 3]
```diff
@@ -1084,3 +1084,11 @@ func (b *BTCMarkets) GetHistoricCandlesExtended(ctx context.Context, p currency.
 	ret.SortCandlesByTimestamp(false)
 	return ret, nil
 }
+
+// GetServerTime returns the current exchange server time.
+func (b *BTCMarkets) GetServerTime(ctx context.Context, ai asset.Item) (time.Time, error) {
+	if !ai.IsValid() {
+		return time.Time{}, fmt.Errorf("%s %w", ai, asset.ErrNotSupported)
+	}
```
Comment (gloriousCode) [inline exchanges/btcmarkets/btcmarkets_wrapper.go:1092]: I don't think this is needed for any that don't distinguish between asset type

File: exchanges/kraken/kraken_wrapper.go [hunk 4]
```diff
@@ -1565,3 +1565,16 @@ func (k *Kraken) GetAvailableTransferChains(ctx context.Context, cryptocurrency
...
+	if !ai.IsValid() {
+		return time.Time{}, fmt.Errorf("%s %w", ai, asset.ErrNotSupported)
+	}
+
+	st, err := k.GetCurrentServerTime(ctx)
+	if err != nil {
+		return time.Time{}, err
+	}
+	return time.Parse("Mon, 02 Jan 06 15:04:05 -0700", st.Rfc1123)
+}
```
Comment (gloriousCode) [inline exchanges/kraken/kraken_wrapper.go:1580]: Please also include Yobit via `GetInfo(ctx context.Context) (Info, error) {`

**Review by gloriousCode — CHANGES_REQUESTED** [review]
This is one of the things I'm surprised we didn't already have. While I think more exchanges actually have this API feature, I'm fine with only adding the currently supported ones at present.

File: exchanges/coinbasepro/coinbasepro_wrapper.go [hunk 5]
```diff
@@ -1019,3 +1019,16 @@ func (c *CoinbasePro) ValidateCredentials(ctx context.Context, assetType asset.I
 	_, err := c.UpdateAccountInfo(ctx, assetType)
 	return c.CheckTransientError(err)
 }
+
+// GetServerTime returns the current exchange server time.
+func (c *CoinbasePro) GetServerTime(ctx context.Context, ai asset.Item) (time.Time, error) {
+	if !ai.IsValid() {
+		return time.Time{}, fmt.Errorf("%s %w", ai, asset.ErrNotSupported)
+	}
```
Comment (gloriousCode) [inline exchanges/coinbasepro/coinbasepro_wrapper.go:1027]: rm

File: exchanges/coinbasepro/coinbasepro_test.go [hunk 6]
```diff
@@ -144,13 +144,30 @@ func TestGetCurrencies(t *testing.T) {
...
 		t.Error("GetServerTime() error", err)
 	}
 }
 
+func TestWrapperGetServerTime(t *testing.T) {
+	t.Parallel()
+	_, err := c.GetServerTime(context.Background(), asset.Empty)
+	if !errors.Is(err, asset.ErrNotSupported) {
+		t.Fatalf("received: '%v' but expected: '%v'", err, asset.ErrNotSupported)
+	}
```
Comment (gloriousCode) [inline exchanges/coinbasepro/coinbasepro_test.go:159]: rm

File: exchanges/btse/btse_wrapper.go [hunk 7]
```diff
@@ -1106,3 +1106,16 @@ func OrderSizeLimits(pair string) (limits OrderSizeLimit, found bool) {
 	val, ok := resp.(OrderSizeLimit)
 	return val, ok
 }
+
+// GetServerTime returns the current exchange server time.
+func (b *BTSE) GetServerTime(ctx context.Context, ai asset.Item) (time.Time, error) {
+	if !ai.IsValid() {
+		return time.Time{}, fmt.Errorf("%s %w", ai, asset.ErrNotSupported)
+	}
```
Comment (gloriousCode) [inline exchanges/btse/btse_wrapper.go:1114]: rm

File: exchanges/btse/btse_test.go [hunk 8]
```diff
@@ -291,14 +291,31 @@ func TestUpdateTickers(t *testing.T) {
...
 		t.Error(err)
 	}
 }
 
+func TestWrapperGetServerTime(t *testing.T) {
+	t.Parallel()
+	_, err := b.GetServerTime(context.Background(), asset.Empty)
+	if !errors.Is(err, asset.ErrNotSupported) {
+		t.Fatalf("received: '%v' but expected: '%v'", err, asset.ErrNotSupported)
+	}
```
Comment (gloriousCode) [inline exchanges/btse/btse_test.go:307]: rm

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for making those changes!

**Review by thrasher- — CHANGES_REQUESTED** [review]
Please add support for Huobi:

`h.GetTimestamp`

And support wrapper additional testing via exchange_wrapper_issues tool

**Review by gloriousCode — APPROVED** [review]
tA:CK

**Review by thrasher- — APPROVED** [review]
Thanks for making those changes!

**Pull request merged**